src/libraries/System.Threading.Channels/tests/BoundedChannelTests.cs

```diff
@@ -249,6 +262,94 @@ public void WriteAsync_TryRead_Many_Ignore(int bufferedCapacity)
+
+            Assert.True(c.Writer.TryWrite(5));
+            Assert.True(c.Writer.TryWrite(5));
+        }
+
+        [Theory]
+        [MemberData(nameof(ChannelDropModes))]
+        public async Task DroppedDelegateIsNull_AsyncWrites(BoundedChannelFullMode boundedChannelFullMode)
+        {
+            Channel<int> c = Channel.CreateBounded<int>(new BoundedChannelOptions(1) { FullMode = boundedChannelFullMode }, itemDropped: null);
+
+            await c.Writer.WriteAsync(5);
+            await c.Writer.WriteAsync(5);
+        }
+
+        [Theory]
+        [MemberData(nameof(ChannelDropModes))]
+        public void DroppedDelegateCalledOnChannelFull_SyncWrites(BoundedChannelFullMode boundedChannelFullMode)
+        {
+            var droppedItems = new HashSet<int>();
+
+            void AddDroppedItem(int itemDropped)
+            {
+                Assert.True(droppedItems.Add(itemDropped));
+            }
+
+            const int channelCapacity = 10;
+            var c = Channel.CreateBounded<int>(new BoundedChannelOptions(channelCapacity)
+            {
+                FullMode = boundedChannelFullMode
+            }, AddDroppedItem);
+
+            for (int i = 0; i < channelCapacity; i++)
+            {
+                Assert.True(c.Writer.TryWrite(i));
+            }
+
+            // No dropped delegate should be called while channel is not full
+            Assert.Empty(droppedItems);
+
+            for (int i = channelCapacity; i < channelCapacity + 10; i++)
+            {
+                Assert.True(c.Writer.TryWrite(i));
+            }
+
+            // Assert expected number of dropped items delegate calls
+            Assert.Equal(10, droppedItems.Count);
+        }
+
+        [Theory]
+        [MemberData(nameof(ChannelDropModes))]
+        public async Task DroppedDelegateCalledOnChannelFull_AsyncWrites(BoundedChannelFullMode boundedChannelFullMode)
+        {
+            var droppedItems = new HashSet<int>();
+
+            void AddDroppedItem(int itemDropped)
+            {
+                Assert.True(droppedItems.Add(itemDropped));
+            }
+
+            const int channelCapacity = 10;
+            var c = Channel.CreateBounded<int>(new BoundedChannelOptions(channelCapacity)
+            {
+                FullMode = boundedChannelFullMode
+            }, AddDroppedItem);
+
+            for (int i = 0; i < channelCapacity; i++)
+            {
+                await c.Writer.WriteAsync(i);
+            }
+
+            // No dropped delegate should be called while channel is not full
+            Assert.Empty(droppedItems);
+
+            for (int i = channelCapacity; i < channelCapacity + 10; i++)
+            {
+                await c.Writer.WriteAsync(i);
+            }
+
+            // Assert expected number of dropped items delegate calls
+            Assert.Equal(10, droppedItems.Count);
+        }
+
```

Comment:
Can you add a test that validates the delegate isn't invoked when BoundedChannelFullMode.Wait is used?

Reply:
Also, a test that helps validate the lock isn't held while invoking the callback would be good, e.g. by having the callback do something like:
```C#
var mres = new ManualResetEventSlim();
ThreadPool.QueueUserWorkItem(delegate
{
    c.Writer.TryWrite(...);
    mres.Set();
});
mres.Wait();
```
If the lock was held while invoking the callback, code like the above should deadlock.

Reply:
Added both test cases.